generation/generate-test.go

```diff
@@ -0,0 +1,335 @@
+// Generates a full netsim test, given a log-splicer processed ssb-fixtures folder and a replication expectations file
+// expectations.json.
+package generation
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"math/rand"
+	"os"
+	"path"
+	"sort"
+)
+
+type Args struct {
+	SSBServer        string
+	FixturesRoot     string
+	ExpectationsPath string
+	FocusedCount     int
+	MaxHops          int
+}
+
+type Generator struct {
+	FocusGroup         []string
+	IDsToNames         map[string]string
+	NamesToIDs         map[string]string
+	currentlyExecuting map[string]bool
+	isBlocking         map[string]map[string]bool
+	Args               Args
+}
+
+func check(err error) {
+	if err != nil {
+		log.Fatalln(err)
+	}
+}
+
+func PickName(splicedFixturesMap map[string]interface{}) string {
+	return splicedFixturesMap["folder"].(string)
+}
+
+// Returns a map of follows (id -> slice of ids that are followed), a map of blocks (g.isBlocking[id][otherId] = true if id blocks otherId)
+func GetFollowMap(followGraphPath string) (map[string][]string, map[string]map[string]bool, error) {
+	// get the json map of all known relations
+	b, err := os.ReadFile(followGraphPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// unpack into goland
+	var v map[string]map[string]interface{}
+	err = json.Unmarshal(b, &v)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	followMap := make(map[string][]string)
+	blockMap := make(map[string]map[string]bool)
+	for id, relations := range v {
+		var following []string
+		blockMap[id] = make(map[string]bool)
+		for relationId, status := range relations {
+			if isFollowing, ok := status.(bool); ok {
+				if isFollowing {
+					following = append(following, relationId)
+				} else {
+					blockMap[id][relationId] = true
+				}
+			}
+			followMap[id] = following
+		}
+	}
+	return followMap, blockMap, nil
+}
+
+func GetIdentities(fixturesRoot string) (map[string]string, error) {
+	b, err := os.ReadFile(path.Join(fixturesRoot, "secret-ids.json"))
+	if err != nil {
+		return nil, err
```

Comment:
see above

Reply:
by which you mean: comment the meaning of `map[string]string`?

Reply:
oh, right.. that too!

actually meant the errors... 😬

Reply:
ahh okay, yeah haha. glad i clarified!